Process swap on cross-site window.open behind a flag

pascoej opened the pull request:
#### f2c0fb94999f81c0383bbb80e186a983962e42c6
<pre>
Process swap on cross-site window.open behind a flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=252339">https://bugs.webkit.org/show_bug.cgi?id=252339</a>
rdar://problem/105511244

Reviewed by NOBODY (OOPS!).

This patch starts process swapping upon cross-site navigations in a window opened via
window.open with an opener relationship. In order to accomplish this, we pass along
the opener to the UI process when creating a page. Then, whenever a cross-site navigation
that causes a process swap occurs, the UI process notifies both the opener process and openee
process of their relationship.

This is accomplished by sending pages with remote main frames to each process and including
info about them whenever creating the process swapped page as well as the page with a remote
main frame in the opener&apos;s process.

The above functionality is guarded by the processSwapsOnWindowOpenWithOpener option on a WebProcessPool.
This is exposed as a feature flag, off by default, as &quot;Swap Processes on Cross-Site Window Open.&quot;

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WindowProxy.cpp:
(WebCore::WindowProxy::replaceFrame):
* Source/WebCore/bindings/js/WindowProxy.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fallbackBaseURL const):
(WebCore::Document::canNavigateInternal):
(WebCore::Document::initSecurityContext):
(WebCore::Document::initContentSecurityPolicy):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::doCrossOriginOpenerHandlingOfResponse):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::detachFromAllOpenedFrames):
(WebCore::FrameLoader::opener):
(WebCore::FrameLoader::opener const):
(WebCore::FrameLoader::setOpener):
(WebCore::FrameLoader::setOriginalURLForDownloadRequest):
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
(WebCore::FrameLoader::effectiveReferrerPolicy const):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::resetWindowProxy):
(WebCore::Frame::takeWindowProxy):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/Platform/IPC/MessageReceiverMap.cpp:
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
(WebKit::WebPageCreationParameters::SubframeProcessFrameTreeInitializationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/OpenedPageProxy.cpp: Added.
(WebKit::OpenedPageProxy::OpenedPageProxy):
(WebKit::OpenedPageProxy::~OpenedPageProxy):
(WebKit::OpenedPageProxy::messageSenderConnection const):
(WebKit::OpenedPageProxy::messageSenderDestinationID const):
(WebKit::OpenedPageProxy::didReceiveMessage):
(WebKit::OpenedPageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/OpenedPageProxy.h: Added.
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::takeOpenedPage):
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didCreateMainFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
(WebKit::ProvisionalPageProxy::openedPage const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::swapToProvisionalPage):
(WebKit::WebPageProxy::close):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::swapToProvisionalPage): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createWebPage):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
(WebKit::WebResourceLoadObserver::logUserInteractionWithReducedTimeResolution):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::constructFrameTree):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::createWebPage):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19fa706191569de19f3b2caf69d536244bb83c1f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6854 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8229 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6920 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6637 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8017 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9814 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6756 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6074 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8316 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4809 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6004 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13840 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5605 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6592 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6084 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8763 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6226 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5385 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6775 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5968 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1572 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10140 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6958 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6341 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1701 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -315,7 +315,11 @@ JSValue toJS(JSGlobalObject* lexicalGlobalObject, DOMWindow& domWindow)
 JSDOMWindow* toJSDOMWindow(Frame& frame, DOMWrapperWorld& world)
 {
-    return frame.script().globalObject(world);
+    if (auto* window = JSC::jsDynamicCast<JSDOMWindow*>(frame.script().globalObject(world))) {
```
In which case isn't the global object of a Frame a `JSDOMWindow*`?

I have concerns this shouldn't be possible and may negatively impact performance.

Note that remote windows should have remote frames. Here we have a (Local)Frame so I don't really understand why its global object wouldn't be a JSDOMWindow.

cdumez commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -315,7 +315,11 @@ JSValue toJS(JSGlobalObject* lexicalGlobalObject, DOMWindow& domWindow)
+    if (auto* window = JSC::jsDynamicCast<JSDOMWindow*>(frame.script().globalObject(world))) {
+        return window;
+    }
+    ASSERT_NOT_REACHED();
```
So this should not be possible, yet we're paying a runtime cost now with the dynamic cast. This doesn't seem ideal.

cdumez commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -58,20 +60,31 @@ WindowProxy::WindowProxy(AbstractFrame& frame)
 {
 }
 
+WindowProxy::WindowProxy(RefPtr<RemoteDOMWindow> window)
```
`RefPtr<RemoteDOMWindow>&&`

cdumez commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -58,20 +60,31 @@ WindowProxy::WindowProxy(AbstractFrame& frame)
 }
 
+WindowProxy::WindowProxy(RefPtr<RemoteDOMWindow> window)
+    : m_remoteWindow(window)
```
`WTFMove(window)`

cdumez commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -58,20 +60,31 @@ WindowProxy::WindowProxy(AbstractFrame& frame)
     ASSERT(m_jsWindowProxies->isEmpty());
 }
 
+Ref<WindowProxy> WindowProxy::create(RefPtr<RemoteDOMWindow> window)
```
`RefPtr<RemoteDOMWindow>&&`

cdumez commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -58,20 +60,31 @@ WindowProxy::WindowProxy(AbstractFrame& frame)
 
+Ref<WindowProxy> WindowProxy::create(RefPtr<RemoteDOMWindow> window)
+{
+    return adoptRef(*new WindowProxy(window));
```
`WTFMove(window)`

cdumez commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -58,20 +60,31 @@ WindowProxy::WindowProxy(AbstractFrame& frame)
 void WindowProxy::detachFromFrame()
 {
-    ASSERT(m_frame);
+    ASSERT(m_frame || m_remoteWindow);
```
I don't really get this. A remote window should have a remote frame and thus shouldn't require special handling?

cdumez commented on Source/WebCore/bindings/js/WindowProxy.cpp:
```diff
@@ -95,14 +108,14 @@ void WindowProxy::destroyJSWindowProxy(DOMWrapperWorld& world)
 JSWindowProxy& WindowProxy::createJSWindowProxy(DOMWrapperWorld& world)
 {
-    ASSERT(m_frame);
+    ASSERT(m_remoteWindow || m_frame);
```
Ditto.

cdumez commented on Source/WebCore/bindings/js/WindowProxy.h:
```diff
@@ -91,11 +94,13 @@ class WindowProxy : public RefCounted<WindowProxy> {
     WEBCORE_EXPORT JSWindowProxy& createJSWindowProxyWithInitializedScript(DOMWrapperWorld&);
 
     WeakPtr<AbstractFrame> m_frame;
+    RefPtr<RemoteDOMWindow> m_remoteWindow;
```
This doesn't look right. We should be getting the remote window from m_frame (which would be a RemoteFrame).

cdumez commented on Source/WebCore/bindings/js/WindowProxy.h:
```diff
@@ -91,11 +94,13 @@ class WindowProxy : public RefCounted<WindowProxy> {
 
 private:
     explicit WindowProxy(AbstractFrame&);
+    explicit WindowProxy(RefPtr<RemoteDOMWindow>);
```
Shouldn't be needed.

cdumez commented on Source/WebCore/page/DOMWindow.h:
```diff
@@ -490,6 +498,8 @@ class DOMWindow final
 
     std::optional<ReducedResolutionSeconds> m_frozenNowTimestamp;
 
+    mutable RefPtr<WindowProxy> m_remoteOpener;
```
I don't think this doesn't belong here. You can get your opener from the FrameLoader. It is should be a RemoteFrame when remote and you should get the window from that.

pascoej replied:
I wanted to use a RemoteFrame, but got guidance not to do that. I will sync up about it.

cdumez replied:
Well, I don't know who gave this guidance but I don't think you can implement it this way. Despite the name, a WindowProxy actually maps closely to a Frame, not to a DOMWindow. Let me explain why I say this:
1. I am on google.com and I call `popup = window.open("https://www.yahoo.com")`
2. Initially, popup is a WindowProxy which points to an about:blank window (in process)
3. The popup gets navigated asynchronously to https://www.yahoo.com, which causes a process-swap.
4. So now `popup` is a WindowProxy which points to the yahoo.com window in another process.
5. Now let's say that the popup gets navigated to "google.com"
6. Your popup should is not a WindowProxy pointing to a "google.com" window that is in your process and which you can script directly. Note that the "google.com" window here is a completely new window (distinct from the "yahoo.com" one)

As you can see, a WindowProxy can point to several window objects since the window object changes whenever the frame gets navigated. The constant for the WindowProxy is the frame is points too and whenever the frame gets navigated, the WindowProxy (initially returned by `window.open()`) will forward calls to a different window, in the same frame. The window/frame can be remote or local. It can go from local to remote and vice versa.

cdumez commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -41,6 +41,13 @@ RemoteDOMWindow::RemoteDOMWindow(RemoteFrame& frame, GlobalWindowIdentifier&& id
 
+RemoteDOMWindow::RemoteDOMWindow(GlobalWindowIdentifier&& identifier, UniqueRef<RemoteDOMWindowClient>&& client)
+    : AbstractDOMWindow(WTFMove(identifier))
+    , m_frame(nullptr)
```
Not needed.

cdumez commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -80,7 +87,7 @@ void RemoteDOMWindow::blur()
 {
     // FIXME: Implemented this.
-    return 0;
+    return 5;
```
Why do this? This can break iteration.

cdumez commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -63,7 +70,7 @@ void RemoteDOMWindow::close(Document&)
 bool RemoteDOMWindow::closed() const
 {
-    return !m_frame;
+    return false;
```
Why?

cdumez reviewed (changes requested):
Only a partial review because the design doesn't seem right.

cdumez commented on Source/WebCore/page/RemoteDOMWindow.h:
```diff
@@ -72,11 +80,14 @@ class RemoteDOMWindow final : public AbstractDOMWindow {
 
     WeakPtr<RemoteFrame> m_frame;
+    WeakPtr<WindowProxy> m_windowProxy;
+    std::optional<UniqueRef<RemoteDOMWindowClient>> m_client;
```
an optional UniqueRef should be a unique_ptr.

cdumez commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -481,6 +481,8 @@ void ProvisionalPageProxy::didReceiveMessage(IPC::Connection& connection, IPC::D
     if (decoder.messageName() == Messages::WebPageProxy::DidStartProgress::name()
         || decoder.messageName() == Messages::WebPageProxy::DidChangeProgress::name()
+        || decoder.messageName() == Messages::WebPageProxy::DidCreateWindow::name()
+        || decoder.messageName() == Messages::WebPageProxy::RemotePostMessage::name()
```
Why this change? Seems odd to expect such message from a provisional page.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -331,8 +331,13 @@ Page* WebChromeClient::createWindow(Frame& frame, const WindowFeatures& windowFe
     WebFrame* webFrame = WebFrame::fromCoreFrame(frame);
+    
+    std::optional<WebCore::GlobalWindowIdentifier> openerWindowID;
+    if (auto *window = frame.window()) {
```
star on wrong side.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -331,8 +331,13 @@ Page* WebChromeClient::createWindow(Frame& frame, const WindowFeatures& windowFe
+    
+    std::optional<WebCore::GlobalWindowIdentifier> openerWindowID;
+    if (auto *window = frame.window()) {
+        openerWindowID = window->identifier();
```
I don't think this is correct in the case where "noopener" is specified. Should we really always be passing an openerWindowID even when `noopener` was specified and `popup.opener` will be null?

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebRemoteDOMWindowClient.h:
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
why 2022?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1699,6 +1701,7 @@ void WebPage::suspendForProcessSwap()
     // Back/forward cache does not break the opener link for the main frame (only does so for the subframes) because the
     // main frame is normally re-used for the navigation. However, in the case of process-swapping, the main frame
     // is now hosted in another process and the one in this process is in the cache.
+    // Maybe this is improtant
```
Probably didn't mean to keep this comment? (+typo)

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -49,6 +49,8 @@
 #include "Logging.h"
 #include "MediaKeySystemPermissionRequestManager.h"
 #include "MediaRecorderProvider.h"
+#include "WebRemoteFrameClient.h"
```
Why these includes?

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -960,6 +964,78 @@ void WebProcess::didReceiveMessage(IPC::Connection& connection, IPC::Decoder& de
 
+void WebProcess::setRemoteOpener(WebCore::GlobalWindowIdentifier&& remoteWindowIdentifier, WebCore::GlobalWindowIdentifier&& openerWindowIdentifier)
+{
+    auto it = WebCore::AbstractDOMWindow::allWindows().find(remoteWindowIdentifier);
```
Seems we could use `get()`. Let's not use iterators unless there is a performance benefit (avoid double hash lookup). Same comment applies below.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4784,6 +4785,46 @@ void WebPageProxy::disconnectFramesFromPage()
+    if (!m_mainFrame || m_mainFrame->frameID() != frameID)
+        return;
+    if (!m_initialWindow)
+        m_initialWindow = windowIdentifier;
```
Assuming initial window means "main frame window" here, I think we should rename to "m_mainFrameWindow" for clarity.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4784,6 +4785,46 @@ void WebPageProxy::disconnectFramesFromPage()
+    if (openerID)
+        m_opener = openerID;
+
+    if (m_opener && m_initialWindow && windowIdentifier.processIdentifier != m_opener->processIdentifier) {
```
I understand you only care about window openers in this patch but I think we should try and be a little bit more forward-looking and try to be a little bit more generic. This code seems very specific to opener and it is not clear to me why it needs to be.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1061,7 +1061,8 @@ bool FrameLoader::checkIfFormActionAllowedByCSP(const URL& url, bool didReceiveR
 {
-    if (m_opener && !opener)
+    auto* window = m_frame.window();
+    if ((m_opener && !opener) || (window && window->hasRemoteOpener()))
```
Don't we need to let the UIProcess know when the main frame's opener changes? Since it is being stored on WebPageProxy?

cdumez commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -113,13 +120,37 @@ WindowProxy* RemoteDOMWindow::parent() const
+        // It doesn't make sense target a postMessage at an opaque origin
+        // because there's no way to represent an opaque origin in a string.
+        if (target->isOpaque())
+            return; //Exception { SyntaxError };
```
If this should return an exception, why doesn't it?

achristensen07 commented on Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
```diff
@@ -177,7 +177,7 @@ class ProcessPoolConfiguration final : public ObjectImpl<Object::Type::ProcessPo
     bool m_processSwapsOnNavigationFromExperimentalFeatures { false };
     bool m_alwaysKeepAndReuseSwappedProcesses { false };
-    bool m_processSwapsOnWindowOpenWithOpener { false };
+    bool m_processSwapsOnWindowOpenWithOpener { true };
```
This should probably get its value from the yaml instead of just switching it to true.

achristensen07 commented on Source/WebKit/UIProcess/OpenedPageProxy.cpp:
```diff
@@ -0,0 +1,89 @@
+}
+
+OpenedPageProxy::~OpenedPageProxy()
+{
```
= default;

achristensen07 commented on Source/WebKit/UIProcess/OpenedPageProxy.cpp:
```diff
@@ -0,0 +1,89 @@
+
+void OpenedPageProxy::didReceiveMessage(IPC::Connection& connection, IPC::Decoder& decoder)
+{
+    // FIXME: Removing this will be necessary to getting layout tests to work with site isolation.
```
remove comment.  It is incorrect.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -141,6 +142,11 @@ std::unique_ptr<DrawingAreaProxy> ProvisionalPageProxy::takeDrawingArea()
 
+std::unique_ptr<OpenedPageProxy> ProvisionalPageProxy::takeOpenedPage()
+{
+    return WTFMove(m_openedPage);
```
std::exchange

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -240,13 +256,31 @@ void ProvisionalPageProxy::didCreateMainFrame(FrameIdentifier frameID)
+
+            parameters.subframeProcessFrameTreeInitializationParameters = { {
+                frameID,
+                *frameTreeCreationParameters,
```
WTFMove

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -240,13 +256,31 @@ void ProvisionalPageProxy::didCreateMainFrame(FrameIdentifier frameID)
+            parameters.previousFrameIdentifier = previousMainFrame->frameID();
+
+            auto frameTreeCreationParameters = m_page.frameTreeCreationParameters();
+            frameTreeCreationParameters->frameID = frameID;
```
doctoring up these values after the return of m_page.frameTreeCreationParameters() seems strange.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -240,13 +256,31 @@ void ProvisionalPageProxy::didCreateMainFrame(FrameIdentifier frameID)
+
+    if (auto* opener = m_page.openerFrame()) {
+        if (auto* openerPage = opener->page()) {
+            auto parameters = m_page.creationParameters(m_page.process(), *m_drawingArea, API::WebsitePolicies::create());
```
Can we get the website policies from somewhere rather than just making an empty one?

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1000,7 +1001,7 @@ bool WebPageProxy::shouldUseBackForwardCache() const
 }
 
-void WebPageProxy::swapToProvisionalPage(std::unique_ptr<ProvisionalPageProxy> provisionalPage)
+void WebPageProxy:: swapToProvisionalPage(std::unique_ptr<ProvisionalPageProxy> provisionalPage)
```
remove extra space

achristensen07 commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -494,7 +494,7 @@ void WebLoaderStrategy::scheduleLoadFromNetworkProcess(ResourceLoader& resourceL
     if (frame)
         loadParameters.effectiveSandboxFlags = frame->loader().effectiveSandboxFlags();
-    if (auto openerFrame = frame ? frame->loader().opener() : nullptr) {
+    if (auto* openerFrame = frame ? dynamicDowncast<LocalFrame>(frame->loader().opener()) : nullptr) {
```
This can easily be put into a separate PR that obviously doesn't change behavior.

pascoej replied:
👍 https://github.com/WebKit/WebKit/pull/12989

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -672,6 +672,10 @@ WebPage::WebPage(PageIdentifier pageID, WebPageCreationParameters&& parameters)
     pageConfiguration.performanceLoggingClient = makeUnique<WebPerformanceLoggingClient>(*this);
     pageConfiguration.screenOrientationManager = m_screenOrientationManager.get();
+    if (parameters.openerFrameIdentifier)
+        pageConfiguration.openerFrameIdentifier = *parameters.openerFrameIdentifier;
```
if and operator* can be removed together.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -781,6 +785,17 @@ WebPage::WebPage(PageIdentifier pageID, WebPageCreationParameters&& parameters)
+    if (parameters.isProcessSwap) {
+        if (parameters.openerFrameIdentifier) {
+            auto* openerFrame = WebProcess::singleton().webFrame(*parameters.openerFrameIdentifier);
+            if (openerFrame)
```
Can be combined with previous line.

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -848,7 +848,7 @@ WebPage* WebProcess::webPage(PageIdentifier pageID) const
 {
     // It is necessary to check for page existence here since during a window.open() (or targeted
-    // link) the WebPage gets created both in the synchronous handler and through the normal way. 
+    // link) the WebPage gets created both in the synchronous handler and through the normal way.
```
This change can be omitted.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -559,6 +559,35 @@ function log(msg)
+)PSONRESOURCE";
+
+
+ static const char* windowOpenWithNameCrossSiteOpenerPopupTestBytes = R"PSONRESOURCE(
```
indentation

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -616,6 +645,51 @@ static void runBasicTest(SchemeHandlerShouldBeAsync schemeHandlerShouldBeAsync)
+    EXPECT_TRUE(!!pid2);
+
+    EXPECT_NE(pid1, pid2);
+    TestWebKitAPI::Util::run(&didReceiveAlert);
```
It might be worth using WKWebView._frames to inspect your frame tree.

pascoej replied:
It seems the opener<->openee relationship doesn't change the results of WKWebView._frames both with this feature turned off and turned on.

achristensen07 replied:
Why?  What would it take to make _frames give useful information for this test?

cdumez commented on Source/WebKit/UIProcess/OpenedPageProxy.h:
```diff
@@ -0,0 +1,63 @@
+class WebPageProxy;
+class WebProcessProxy;
+
+class OpenedPageProxy : public IPC::MessageReceiver, public IPC::MessageSender {
```
There is something fishy about the design here. Why do we need an OpenedPageProxy?

When we create a popup, we get a regular WebPageProxy. Then, when the initial navigation is cross-origin, we'll create a ProvisionalPageProxy for the process-swap until the load is committed. We should not need a new `OpenedPageProxy` type.

pascoej replied:
Somewhat analogous to SubframePageProxy to track the remote page in another process. Though as it stands now, it's just light sugar over constructing and sending the CreateWebPage message. I'm fine copy pastaing the contents of OpenedPageProxy's constructor to initializeWebPage for now.

cdumez replied:
`SubframePageProxy` is a bit weird too but I believe this was needed because frames in remote processes needed a WebPage on the WebProcess side and these needed an object to communicate to on the UIProcess side. It couldn't be the WebPageProxy because it is already connected to the WebPage associated with the main frame process.

In your case though, it is a popup. A popup gets a brand new WebPageProxy and this WebPageProxy is associated with a single WebProcess (which may change on navigation). This is just regular process-swap-on-navigation and I stand by my assessment that we shouldn't need a new proxy object on the UIProcess side. Why can't we use that WebPageProxy, why would we need yet another proxy object?

So far, your comments didn't convince me we NEED a new proxy object.

cdumez reviewed (changes requested):
r- because of OpenedPageProxy.

cdumez commented on Source/WebKit/Shared/WebPageCreationParameters.h:
```diff
@@ -292,12 +292,14 @@ struct WebPageCreationParameters {
     };
     std::optional<SubframeProcessFrameTreeInitializationParameters> subframeProcessFrameTreeInitializationParameters;
+    std::optional<WebCore::FrameIdentifier> openerFrameIdentifier;
+    std::optional<WebCore::FrameIdentifier> previousFrameIdentifier;
```
I find `previousFrameIdentifier` to be confusing. I am not sure what it means.

achristensen07 replied:
Ditto.  It seems like an indication that we're changing frame identifiers somewhere, which we shouldn't do.

cdumez commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -170,6 +170,23 @@ void ProvisionalPageProxy::initializeWebPage(RefPtr<API::WebsitePolicies>&& webs
     parameters.isProcessSwap = true;
+
+    auto* openerFrame = m_page->openerFrame();
+    if (openerFrame) {
```
`if (auto* openerFrame = m_page->openerFrame()) {`

cdumez commented on Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
```diff
@@ -170,6 +170,23 @@ void ProvisionalPageProxy::initializeWebPage(RefPtr<API::WebsitePolicies>&& webs
+                std::nullopt,
+            } };
+
+            m_process->send(Messages::WebProcess::CreateWebPage(openerPage->webPageID(), parameters), 0);
```
I am unclear why we need to create a page for the opener in the new process. Wouldn't it suffice to have a RemoteFrame for the opener?

pascoej replied:
A RemoteFrame requires a Page and in this situation it wouldn't be correct to claim that the opened page is the associated page for the frame.

achristensen07 replied:
Is there a UIProcess object that represents this page?

cdumez replied:
If we really need this page in the process (Apparently Alex didn't say otherwise), and if we need a UIProcess-side object to represent it, I recommend we use the same object that Alex already introduced for remote pages (Please let's not add an OpenerPageProxy like the earlier patch iteration). I don't remember what it is called.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -765,17 +767,17 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
     void setViewNeedsDisplay(const WebCore::Region&);
     void requestScroll(const WebCore::FloatPoint& scrollPosition, const WebCore::IntPoint& scrollOrigin, WebCore::ScrollIsAnimated);
-    
```
Plenty of unrelated spacing changes, please revert those.

achristensen07 commented on Source/WebCore/page/Frame.cpp:
```diff
@@ -73,7 +73,8 @@ bool Frame::isRootFrame() const
 void Frame::resetWindowProxy()
 {
-    m_windowProxy->detachFromFrame();
+    if (m_windowProxy->frame() == this)
```
I feel like we ought to just add ASSERT(m_windowProxy->frame() == this) here, otherwise we've already done something wrong.

achristensen07 commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
```diff
@@ -84,18 +87,18 @@ class RemoteLayerTreeEventDispatcher : public ThreadSafeRefCounted<RemoteLayerTr
 
 private:
-    OptionSet<WheelEventProcessingSteps> determineWheelEventProcessing(const WebCore::PlatformWheelEvent&, WebCore::RectEdges<bool> rubberBandableEdges);
+    OptionSet<WebCore::WheelEventProcessingSteps> determineWheelEventProcessing(const WebCore::PlatformWheelEvent&, WebCore::RectEdges<bool> rubberBandableEdges);
```
This change can be done in a separate PR.

pascoej replied:
I think this was needed because OpenedPageProxy changed the unified build order, but shouldn't be anymore if I don't introduce a new file again.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -461,11 +461,11 @@ void WebFrameProxy::getFrameInfo(CompletionHandler<void(FrameTreeNodeData&&)>&&
     return {
-        m_frameID,
-        m_process->coreProcessIdentifier(),
+        remoteMainFrameIdentifier ? *remoteMainFrameIdentifier : m_frameID,
```
This doesn't look right.  Why are we using a different frame identifier here?
Maybe we should have some remote frame information adjacent to the FrameTreeCreationParameters, but not thrown into it like this.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2304,19 +2306,11 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
-        WebCore::ResourceRequest&&, const String& frameName, uint64_t listenerID);
-    void decidePolicyForResponse(IPC::Connection&, WebCore::FrameIdentifier, FrameInfoData&&, WebCore::PolicyCheckIdentifier, uint64_t navigationID,
-        const WebCore::ResourceResponse&, const WebCore::ResourceRequest&, bool canShowMIMEType, const String& downloadAttribute, uint64_t listenerID);
+    void decidePolicyForNavigationAction(Ref<WebProcessProxy>&&, WebCore::PageIdentifier, WebFrameProxy&, FrameInfoData&&, uint64_t navigationID, NavigationActionData&&, FrameInfoData&& originatingFrameInfo, std::optional<WebPageProxyIdentifier> originatingPageID, const WebCore::ResourceRequest& originalRequest, WebCore::ResourceRequest&&, IPC::FormDataReference&& requestBody, WebCore::ResourceRe ...
```
And this newline change looks unrelated.  Do in a separate PR if it's important.

pascoej committed:
Process swap on cross-site window.open behind a flag
https://bugs.webkit.org/show_bug.cgi?id=252339
rdar://problem/105511244

Reviewed by NOBODY (OOPS!).

This patch starts process swapping upon cross-site navigations in a window opened via
window.open with an opener relationship. In order to accomplish this, we pass along
the opener to the UI process when creating a page. Then, whenever a cross-site navigation
that causes a process swap occurs, the UI process notifies both the opener process and openee
process of their relationship.

This is accomplished by sending pages with a remote main frame to the openee process and sending
a didCommitLoadInAnotherProcess to the opener process.

The above functionality is guarded by the processSwapsOnWindowOpenWithOpener option on a WebProcessPool.
This is exposed as a feature flag, off by default, as "Swap Processes on Cross-Site Window Open."

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WindowProxy.cpp:
(WebCore::WindowProxy::replaceFrame):
* Source/WebCore/bindings/js/WindowProxy.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::suspendActiveDOMObjects):
(WebCore::Document::resumeActiveDOMObjects):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::resetWindowProxy):
(WebCore::Frame::takeWindowProxy):
* Source/WebCore/page/Frame.h:
* Source/WebKit/Shared/LocalFrameCreationParameters.h:
* Source/WebKit/Shared/LocalFrameCreationParameters.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didCreateMainFrame):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/SubframePageProxy.cpp:
(WebKit::SubframePageProxy::SubframePageProxy):
* Source/WebKit/UIProcess/SubframePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigation):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::loadRequestByCreatingNewLocalFrameOrConvertingRemoteFrame):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::updateRendering):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
